enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,255 @@
+
+### Implementation Details/Notes/Constraints
+
+The Insights operator currently runs and gathers the data in a single container. 
+Uploading of new archives, pruning old archives as well as downloading the last Insights analysis report (this is always per archive) is implemented in this one container/binary. 
+When there is a successful upload (to the console.redhat.com) of the archive then there is some short delay (waiting for the processing in console.redhat.com pipelines) and then there is request to get the last Insights analysis report per given cluster. 
+Finally the status section of the `insightsoperator.operator.openshift.io` CR is updated. Note that there is a retry logic in case of failed upload and after some time (and number of attempts) the Insights operator is marked as degraded.
+Introducing a new on demand process/container for running an "extra" gather process raises the question about parallelism. It should be feasible to allow two (or more) parallel gather processes, but it is very important to keep in mind that only one (and that is the latest) Insights analysis report will be available.  
+
+
+### Risks and Mitigations
+
+Risk: Inconsistent Insights analysis results. Assume following situation:
+- periodic gathering is running in the Insights operator (let's call it gather A)
+- user runs a new on demand data gathering (let's call it gather B) e.g 30 seconds after gather A
+- gather A and B finish at essentially the same time
+- the external processing service (in the console.redhat.com) provides only one recent Insights analysis result, and there is no guarantee which (of the two) it is.
+- this can be very confusing, because user running gather B can get result of the gather A (e.g there could be different gatherers enabled, which would produce different results)
+
+Mitigation: This is the very first question in the [Open questions](#open-questions) section, but perhaps reporting the Insights analysis result in one place could slightly mitigate this problem (the status part could tell whether it was a periodic gather or user-triggered gather). 
+Another option would be to disallow the parallel gather execution - if the Insights operator sees a new CR instance (or any other running gather process) then it interrupts any gathering in progress. 
+
+### Drawbacks
+
+- Very frequent data gathering in a cluster can produce a significant load on the processing part in the console.redhat.com and can overwhelm the processing pipeline. 
+Theoretically, we could limit the number of user-trigerred gathering over time, but this problem can happen today when something is deleting the Insights operator pod on a regular basis with appropriate timing (so that data can be uploaded).  
+- When the newly spawned container is failing to upload the data then it can be running for some longer time (couple of hours). 
```

Comment:
Can you report this back into the main operator somehow to make errors obvious?

Reply:
We can extend the status section (to report sucessful/failed upload/download) of the new CRD and then the operator could read it (and derive the clusteroperator conditions from that). Would that make sense? 
What is not clear to me (and it's related to the https://github.com/openshift/enhancements/pull/1279/files#r1021631963) how this would work in the periodical gathering (in separate pod). Would periodical gathering report its status to `insightsoperator.operator.openshift.io` resource? I can imagine we would extend this CR to report about the upload and download and then the clusteroperator conditions could be derived from that?

Reply:
> We can extend the status section (to report sucessful/failed upload/download) of the new CRD and then the operator could read it (and derive the clusteroperator conditions from that). Would that make sense?

I had imagined the operator managing the pods reporting failure directly into the `clusteroperator` status, but maybe there needs to be a proxy for the data as well?

In the periodic case I would imagine the controller driving those evolves to:
- On Cron:
  - Create new `DataGather`
  - Watch `DataGather` resource
  - On `DataGather` status update, copy status to `insightsoperator`
  - ^ repeat a few times
  - Once complete, copy `DataGather` status update to `insightsoperator`
  - Sleep until next cron interval

Reply:
Yes I think you described what I meant and that is basically copying the status from the finsihed `DataGather` to `insightsoperator.operator.openshift.io` CR.

Reply:
I believe that has been clarified and agreed. Thank you.